<a href="https://colab.research.google.com/github/aabhashree-github/python-visualization/blob/main/(GitHub)_CMPS_335_Project_2_lamichhanea.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Project 2

This project is about using cross-validation to explore hyperparameter options for two different types of models. We are using a random forest model and a k-nearest neighbors model.

We are using another dataset with a long history in machine learning, a dataset from the Univerisity of Wisconsin on breast cancer.


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_validate
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import StandardScaler

In [ ]:
# Read in the dataset and specify the id column to be row indices
bcwd = pd.read_csv('/content/drive/My Drive/Hollins Data Science/Datasets/breast_cancer_wisconsin_diagnostic.csv', index_col='id')
# Recode the classes with numbers
bcwd['diagnosis'].replace('M',1, inplace=True)
bcwd['diagnosis'].replace('B',0, inplace=True)
bcwd.head()

diagnosis  radius_mean  ...  symmetry_worst  fractal_dimension_worst
id                                ...                                         
842302            1        17.99  ...          0.4601                  0.11890
842517            1        20.57  ...          0.2750                  0.08902
84300903          1        19.69  ...          0.3613                  0.08758
84348301          1        11.42  ...          0.6638                  0.17300
84358402          1        20.29  ...          0.2364                  0.07678

[5 rows x 31 columns]

In [ ]:
# Create training and test sets
# Using a seed for the random number generator to ensure the split is the same every time
# for the purposes of this project
training_set, test_set = train_test_split(bcwd, test_size=0.20, random_state=42)

# Use all of the features, excluding diagnosis at the beginning
training_features = training_set.iloc[:,1:]
# Targets are at the end
training_targets = training_set.iloc[:,0]

# Pull out the same features and targets from test set
testing_features = test_set.iloc[:,1:]
testing_targets = test_set.iloc[:,0]

# Use this list of measurements to score your cross-validation results
scoring = ['accuracy','precision', 'recall', 'f1']

# Part One

In this part we will train a series of random forest models using cross-validation. We will perform what is called a grid search of two different hyperparameters. A grid search explores different combinations of values across multiple hyperparamters. 

For our purposes, we will explore max_features and max_depth. Max_features specifies how many features each tree has access too. Max_depth limits the growth of the tree. We will explore two values (small and large) for max_features, 4 and 20, and two values of max_depth, 4 and 10. This means a total of 4 models will be created. 

The first model with max_features=4 and max_depth=4 has code ready to go. You will need to create the other three models that have different pairings of hyperparameter values. Finally, pick your model with the highest test_accuracy and evaluate it on the test set (Code is given for this, but you must set the hyperparameters based on your results).


In [ ]:
# Sample code of a random forest cross-validation run

# Create the random forest model with max_features=4 and max_depth=4
rf_model = RandomForestClassifier(criterion='entropy', max_features=4, max_depth=4)

# Cross validation calculating multiple metrics
rf_model.fit(training_features, training_targets)
scores = cross_validate(rf_model, training_features, training_targets, scoring=scoring, return_train_score=True, cv=5)

# Average the runs
results = pd.DataFrame(scores)
results.mean()

fit_time           0.174091
score_time         0.010752
test_accuracy      0.947253
train_accuracy     0.991758
test_precision     0.956665
train_precision    1.000000
test_recall        0.899465
train_recall       0.977821
test_f1            0.926379
train_f1           0.988775
dtype: float64

In [ ]:
# Create the random forest model with max_features=4 and max_depth=10
rf_model2 = RandomForestClassifier(criterion='entropy', max_features=4, max_depth=10)

# Cross validation calculating multiple metrics
rf_model2.fit(training_features, training_targets)
scores2 = cross_validate(rf_model2, training_features, training_targets, scoring=scoring, return_train_score=True, cv=5)

# Average the runs
results2 = pd.DataFrame(scores2)
results2.mean()

fit_time           0.176041
score_time         0.010390
test_accuracy      0.960440
train_accuracy     1.000000
test_precision     0.969521
train_precision    1.000000
test_recall        0.923173
train_recall       1.000000
test_f1            0.944860
train_f1           1.000000
dtype: float64

In [ ]:
# Create the random forest model with max_features=20 and max_depth=4
rf_model3 = RandomForestClassifier(criterion='entropy', max_features=20, max_depth=4)

# Cross validation calculating multiple metrics
rf_model3.fit(training_features, training_targets)
scores3 = cross_validate(rf_model3, training_features, training_targets, scoring=scoring, return_train_score=True, cv=5)

# Average the runs
results3 = pd.DataFrame(scores3)
results3.mean()

fit_time           0.351327
score_time         0.010493
test_accuracy      0.956044
train_accuracy     0.993956
test_precision     0.958173
train_precision    1.000000
test_recall        0.922995
train_recall       0.983736
test_f1            0.939339
train_f1           0.991799
dtype: float64

In [ ]:
# Create the random forest model with max_features=20 and max_depth=10
rf_model4 = RandomForestClassifier(criterion='entropy', max_features=20, max_depth=10)
                                   
# Cross validation calculating multiple metrics
rf_model4.fit(training_features, training_targets)
scores4 = cross_validate(rf_model4, training_features, training_targets, scoring=scoring, return_train_score=True, cv=5)

# Average the runs
results4 = pd.DataFrame(scores4)
results4.mean()

fit_time           0.366364
score_time         0.010591
test_accuracy      0.960440
train_accuracy     1.000000
test_precision     0.952491
train_precision    1.000000
test_recall        0.940998
train_recall       1.000000
test_f1            0.946150
train_f1           1.000000
dtype: float64

In [ ]:
# Sample code evaluating an rf model on the test set
# Customize it as appropriate
rf_model5 = RandomForestClassifier(criterion='entropy', max_features=4, max_depth=10)
rf_model5.fit(training_features, training_targets)

rf_model5.score(testing_features, testing_targets)

0.9649122807017544

## Questions on Part One

1) Which model had the highest accuracy test_accuracy score?

*The random forest models with `max_features = 4` and `max_depth=10` and `max_features = 20` and `max_depth=10` had the highest test_accuracy score of 0.960440.*

2) Which model had the highest test_f1 score?

*The random forest model with `max_features = 20` and `max_depth=10` had the highest test_f1 score of 0.946150.*

3) Which model had the highest fit_time? Why do you think this is?

*The random forest model with `max_features = 20` and `max_depth=10` has the highest fit_time of 0.366364. I think it is because the model has access to 20 features and it takes time to go through those 20 features. The model also has a max depth of 10 which means it takes time for the growth of the tree to a depth of 10.*

4) Which model had the lowest fit_time? Why do you think this is?

*The random forest model with `max_features = 4` and `max_depth=4` has the lowest fit_time of 0.183027. I think this it is because the model has the least number of maximum features to use and also the least number of maximum depths for the tree.*

5) Does the model with the highest fit_time have the best f1 score?

*Yes, the random forest model with `max_features = 20` and `max_depth=10` has both the highest fit_time and the best f1 score.*

6) Did you see any evidence of over-fitting from the training accuracy scores?

*No, I do not see any evidence of overfitting from the training accuracy scores as there is no vast difference between the train_accuracy and test_accuracy scores. They are just off by a few points. The test_accuracy scores are also not that relatively poor which would reflect that the model is not performing better on unseen instances.*

7) How did your best model in terms of test_accuracy score do when evaluated on the testing set using the score method? How different were the results from the test_accuracy score during cross-validation? 

*The best model in terms of test_accuracy score did slightly better when evaluated on the testing set using the score method. The accuracy from the score method came out to be 0.96491 whereas the test_accuracy score during cross validation was 0.960440.*

# Part Two
In part two we will explore using k-nearest neighbors algorithm. First, we will investigate whether scaling the data improves results. Second, we will use cross-validation to set the k value of our model. 

Code is given to create and evaluate a k-nearest neighbors model with K=3 on the original data. Code is then given to scale the data. You need to then recreate the knn model with K=3 and evalute its performance on the scaled data. Then create knn models with K=7 and K=12 and evaluate them. Finally, pick your model with the highest test_accuracy and evaluate it on the test set (Code is given for this, but you must set the correct K value).


In [ ]:
training_set, test_set = train_test_split(bcwd, test_size=0.20, random_state=42)

# Use all of the features, excluding diagnosis at the beginning
training_features = training_set.iloc[:,1:]
# Targets are at the end
training_targets = training_set.iloc[:,0]

# Pull out the same features and targets from test set
testing_features = test_set.iloc[:,1:]
testing_targets = test_set.iloc[:,0]

# Use this list of measurements to score your cross-validation results
scoring = ['accuracy','precision', 'recall', 'f1']

In [ ]:
# Create KNN model with n_neighbors=3, train on original data
knn_model = KNeighborsClassifier(n_neighbors=3)

# Cross validation calculating multiple metrics
knn_model.fit(training_features, training_targets)
scores = cross_validate(knn_model, training_features, training_targets, scoring=scoring, return_train_score=True, cv=5)

# Average the runs
results = pd.DataFrame(scores)
results.mean()

fit_time           0.002778
score_time         0.007360
test_accuracy      0.920879
train_accuracy     0.948901
test_precision     0.934580
train_precision    0.964923
test_recall        0.846702
train_recall       0.895000
test_f1            0.886192
train_f1           0.928605
dtype: float64

In [ ]:
# Scale the features to have a mean of 0 and a standard deviation of 1
feature_scaler = StandardScaler().fit(training_features)
training_features_scaled = feature_scaler.transform(training_features)
testing_features_scaled = feature_scaler.transform(testing_features)

In [ ]:
# Create KNN model with n_neighbors=3, train on scaled data
knn_model_sc = KNeighborsClassifier(n_neighbors=3)

# Cross validation calculating multiple metrics
knn_model_sc.fit(training_features_scaled, training_targets)
scores_sc = cross_validate(knn_model_sc, training_features_scaled, training_targets, scoring=scoring, return_train_score=True, cv=5)

# Average the runs
results_sc = pd.DataFrame(scores_sc)
results_sc.mean() 

fit_time           0.001431
score_time         0.008014
test_accuracy      0.960440
train_accuracy     0.982418
test_precision     0.974531
train_precision    0.996947
test_recall        0.917291
train_recall       0.955621
test_f1            0.944211
train_f1           0.975814
dtype: float64

In [ ]:
# Create KNN model with n_neighbors=7, train on scaled data
knn_model2_sc = KNeighborsClassifier(n_neighbors=7)

# Cross validation calculating multiple metrics
knn_model2_sc.fit(training_features_scaled, training_targets)
scores2_sc = cross_validate(knn_model2_sc, training_features_scaled, training_targets, scoring=scoring, return_train_score=True, cv=5)

# Average the runs
results2_sc = pd.DataFrame(scores2_sc)
results2_sc.mean() 

fit_time           0.001331
score_time         0.007638
test_accuracy      0.960440
train_accuracy     0.971429
test_precision     0.981339
train_precision    0.998374
test_recall        0.911586
train_recall       0.924575
test_f1            0.943823
train_f1           0.959980
dtype: float64

In [ ]:
# Create KNN model with n_neighbors=12, train on scaled data
knn_model3_sc = KNeighborsClassifier(n_neighbors=12)

# Cross validation calculating multiple metrics
knn_model3_sc.fit(training_features_scaled, training_targets)
scores3_sc = cross_validate(knn_model3_sc, training_features_scaled, training_targets, scoring=scoring, return_train_score=True, cv=5)

# Average the runs
results3_sc = pd.DataFrame(scores3_sc)
results3_sc.mean() 

fit_time           0.001443
score_time         0.008831
test_accuracy      0.953846
train_accuracy     0.963187
test_precision     0.993548
train_precision    0.998319
test_recall        0.881818
train_recall       0.902364
test_f1            0.933414
train_f1           0.947875
dtype: float64

In [ ]:
# Sample code evaluating a knn model on the scaled test set
# Customize it as appropriate
knn_model = KNeighborsClassifier(n_neighbors=3)
knn_model.fit(training_features_scaled, training_targets)

# The score method is quick way of getting the accuracy on the testing data
knn_model.score(testing_features_scaled, testing_targets)

0.9473684210526315

## Questions for Part Two

1) Did the results from KNN model change when you switched to scaled features?

*Yes, the results from the KNN model change when I switched to scaled features. All the scores of the model with scaled features improved than those for the model without scaled features.* 

2) Which KNN model had the highest test_accuracy? Did it also have the highest test_f1 score? Is precision or recall higher and what does that say about the model's performance?

*Two models with scaled features, each with n_neighbors = 3 and n_neighbors = 7 had the highest test_accuracy of 0.960440. The scaled model with n_neighbors = 3 had the highest test_f1 score of 0.944211.*

*For the model with n_neighbors = 3, both the test_precision and train_precision are higher than test_recall and train_recall. This tells us that the precentage of total relevant results correctly classified by the model is lower than the actual percentage of relevant results in the model.*

3) How did your best model in terms of test_accuracy do when evaluated on the testing set? Did the accuracy go up or down and by how much?

*My best model in terms of test_accuracy did fine when evaluated on the testing set. The accuracy went slightly down by 1.3072% (from 96.0440% to 94.7368%).*

4) How does your best KNN model compare to your best Random Forest model in terms of accuracy on the testing set?

*My best Random Forest model did better than my best KNN model in terms of accuracy on the testing set. The accuracy on the testing set from the best random forest model was 96.491% (slightly higher than the test_accuracy score during cross validation which was 96.0440%) whereas the accuracy on the testing set model from the best KNN model was 94.7368% (lower that the test_accuracy, 96.0440%).*

5) How do fit_times compare between the KNN models and the Random Forest models?

*The fit_times for the KNN models are much lower than the fit_times for the Random Forest models. This might be because the random forest model has to build a tree in order to reach to the features with the best information gain in order to make its predictions.*

*THANK YOU*